File: golang/graph/graph_test.go
package graph

import (
	"encoding/json"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// BuildComentionGraph is proven against the shared fixture — every case must
// match the Python arbiter (citenexus.graph.store.build_comention_graph)
// exactly. Follows the gate/tokenize exemplar: load the fixture, assert over ALL
// cases, no leniency. Produced and expected are both marshalled to canonical
// JSON (field order + null/[] normalized) and compared byte-for-byte.
func TestComentionGraphConformance(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name     string          `json:"name"`
			Rows     []Row           `json:"rows"`
			Expected json.RawMessage `json:"expected"`
		} `json:"cases"`
	}
	conform.Case(t, "graph_comention.json", &fixture)

	if len(fixture.Cases) == 0 {
		t.Fatal("no graph_comention cases loaded")
	}

	for _, c := range fixture.Cases {
		var want Index
		if err := json.Unmarshal(c.Expected, &want); err != nil {
			t.Fatalf("%s: unmarshal expected: %v", c.Name, err)
		}
		got := BuildComentionGraph(c.Rows)

		gotJSON, err := json.Marshal(got)
		if err != nil {
			t.Fatalf("%s: marshal got: %v", c.Name, err)
		}
		wantJSON, err := json.Marshal(want)
		if err != nil {
			t.Fatalf("%s: marshal want: %v", c.Name, err)
		}
		if string(gotJSON) != string(wantJSON) {
			t.Errorf("%s:\n got  %s\n want %s", c.Name, gotJSON, wantJSON)
		}
	}
}

File: js/src/graph/graph.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { buildComentionGraph, type GraphRow, type GraphIndex } from "./graph.js";

// Proven against the shared fixture — every case must match the Python reference
// citenexus.graph.store.build_comention_graph exactly. Follows the §4 exemplar:
// load conformance/cases/graph_comention.json, assert deep equality over ALL
// cases, no leniency.
interface GraphCase {
  name: string;
  rows: GraphRow[];
  expected: GraphIndex;
}

/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 3 };

describe("co-mention graph conformance", () => {
  const { cases } = loadCase<{ cases: GraphCase[] }>("graph_comention.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const c of cases) {
    it(c.name, () => {
      expect(buildComentionGraph(c.rows)).toEqual(c.expected);
    });
  }
});
